diff --git a/drivers/gpu/drm/msm_default/Makefile b/drivers/gpu/drm/msm_default/Makefile
index 23ce07332e5f..835b06cab4eb 100644
--- a/drivers/gpu/drm/msm_default/Makefile
+++ b/drivers/gpu/drm/msm_default/Makefile
@@ -161,8 +161,7 @@ msm-display-$(CONFIG_DRM_MSM_DSI_7NM_PHY) += dsi/phy/dsi_phy_7nm.o
 
 msm-y += $(adreno-y) $(msm-display-y)
 
-obj-$(CONFIG_DRM_MSM)	+= msm_display.o
-msm_display-y := $(msm-y)
+obj-$(CONFIG_DRM_MSM)	+= msm.o
 
 ifeq (y,$(CONFIG_DRM_MSM_VALIDATE_XML))
 	headergen-opts += --validate
diff --git a/drivers/gpu/drm/msm_default/msm_drv.c b/drivers/gpu/drm/msm_default/msm_drv.c
index 07908b3c56d0..e6727e2cc9a5 100644
--- a/drivers/gpu/drm/msm_default/msm_drv.c
+++ b/drivers/gpu/drm/msm_default/msm_drv.c
@@ -45,6 +45,7 @@
 #define MSM_VERSION_PATCHLEVEL	0
 
 static void msm_deinit_vram(struct drm_device *ddev);
+static const char *mdss_with_gpu_name = NULL;
 
 static char *vram = "16m";
 MODULE_PARM_DESC(vram, "Configure VRAM size (for devices without IOMMU/GPUMMU)");
@@ -215,6 +216,8 @@ static int msm_drm_init(struct device *dev, const struct drm_driver *drv)
 	struct msm_drm_private *priv = dev_get_drvdata(dev);
 	struct drm_device *ddev;
 	int ret;
+	const char *dev_name = NULL;
+	static int card_created = 0;
 
 	if (drm_firmware_drivers_only())
 		return -ENODEV;
@@ -282,10 +285,19 @@ static int msm_drm_init(struct device *dev, const struct drm_driver *drv)
 		ddev->driver_features &= ~DRIVER_ATOMIC;
 	}
 
+	dev_name = of_node_full_name(dev->of_node);
+
+	if (!card_created && mdss_with_gpu_name && strcmp(dev_name, mdss_with_gpu_name)) {
+		ret = -EPROBE_DEFER;
+		goto err_msm_uninit;
+	}
+
 	ret = drm_dev_register(ddev, 0);
 	if (ret)
 		goto err_msm_uninit;
 
+	card_created = 1;
+
 	ret = msm_debugfs_late_init(ddev);
 	if (ret)
 		goto err_msm_uninit;
@@ -1034,6 +1046,7 @@ int msm_drv_probe(struct device *master_dev,
 	struct msm_drm_private *priv;
 	struct component_match *match = NULL;
 	int ret;
+	static int gpu_added = 0;
 
 	priv = devm_kzalloc(master_dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
@@ -1050,6 +1063,14 @@ int msm_drv_probe(struct device *master_dev,
 			return ret;
 	}
 
+	if (!gpu_added) {
+		ret = add_gpu_components(master_dev, &match);
+		if (ret)
+			return ret;
+		gpu_added++;
+		mdss_with_gpu_name = of_node_full_name(master_dev->of_node);
+	}
+
 	/* on all devices that I am aware of, iommu's which can map
 	 * any address the cpu can see are used:
 	 */
@@ -1098,6 +1119,7 @@ static int __init msm_drm_register(void)
 	msm_dsi_register();
 	msm_hdmi_register();
 	msm_dp_register();
+	adreno_register();
 	msm_mdp4_register();
 	msm_mdss_register();
 	return platform_driver_register(&msm_platform_driver);
@@ -1111,6 +1133,7 @@ static void __exit msm_drm_unregister(void)
 	msm_mdp4_unregister();
 	msm_dp_unregister();
 	msm_hdmi_unregister();
+	adreno_unregister();
 	msm_dsi_unregister();
 	msm_mdp_unregister();
 	msm_dpu_unregister();
